msgraphcore/graph_client.py

```diff
@@ -1,124 +1,105 @@
-"""
-Graph Session
-"""
+from typing import List, Optional
+
 from requests import Session
 
-from msgraphcore.constants import BASE_URL, SDK_VERSION
+from msgraphcore.client_factory import HTTPClientFactory
 from msgraphcore.middleware.abc_token_credential import TokenCredential
-from msgraphcore.middleware.authorization import AuthorizationHandler
-from msgraphcore.middleware.middleware import BaseMiddleware, MiddlewarePipeline
+from msgraphcore.middleware.middleware import BaseMiddleware
 from msgraphcore.middleware.options.middleware_control import middleware_control
 
 
-class GraphSession(Session):
-    """Extends Session with Graph functionality
-
-    Extends Session by adding support for middleware options and middleware pipeline
-    """
-    def __init__(
-        self,
-        credential: TokenCredential,
-        scopes: [str] = ['.default'],
-        middleware: list = [],
-        api_version: str = 'v1.0'
-    ):
-        super().__init__()
-        self._append_sdk_version()
-        self._base_url = BASE_URL + '/' + api_version
+class GraphClient:
+    """Constructs a custom HTTPClient to be used for requests against Microsoft Graph"""
+    __instance = None
 
-        auth_handler = AuthorizationHandler(credential, scopes)
+    def __new__(cls, *args, **kwargs):
+        if not GraphClient.__instance:
+            GraphClient.__instance = object.__new__(cls)
+        return GraphClient.__instance
 
-        # The authorization handler should be the first middleware in the pipeline.
-        middleware.insert(0, auth_handler)
-        self._register(middleware)
+    def __init__(self, **kwargs):
+        """
+        Class constructor that accepts a session object and kwargs to
+        be passed to the HTTPClientFactory
+        """
+        self.graph_session = self.get_graph_session(**kwargs)
 
     @middleware_control.get_middleware_options
     def get(self, url: str, **kwargs):
         r"""Sends a GET request. Returns :class:`Response` object.
-
         :param url: URL for the new :class:`Request` object.
         :param \*\*kwargs: Optional arguments that ``request`` takes.
         :rtype: requests.Response
         """
-        return super().get(self._graph_url(url), **kwargs)
+        return self.graph_session.get(self._graph_url(url), **kwargs)
```

Comment:
It will be a better experience for devs to get intellisense for the expected parameters. Using **kwargs as is means devs don't get intellisense.